examples/albef/model.py

```diff
@@ -0,0 +1,357 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+import copy
+from typing import Callable, List, Optional, Tuple, Union
+
+import torch
+import torch.nn.functional as F
+from torch import nn, Tensor
+from torchmultimodal.models.albef import ALBEFModel
+from torchmultimodal.modules.encoders.albef_multimodal_encoder import (
+    ALBEFMultimodalEncoder,
+)
+from torchmultimodal.modules.encoders.albef_text_encoder import (
+    ALBEFTextEmbeddings,
+    ALBEFTextEncoder,
+)
+from torchmultimodal.modules.encoders.albef_vision_encoder import ALBEFVisionEncoder
+from torchmultimodal.modules.losses.albef import MaskedLanguageModelingLoss
+from torchmultimodal.utils.attention import get_extended_attention_mask_for_decoder
+from torchmultimodal.utils.common import momentum_update, remove_grad
+
+
+class PredictionHead(nn.Module):
+    """
+    Predict the following token autoregressively.
+
+    Args:
+        vocab_size (int): The number of different tokens the prediction_head can predict.
+        hidden_size (int): The hidden size of the prediction_head.
+        layer_norm_eps (float): The epsilon used by the prediction_head normalization layer.
+        transform_act_fn (Callable[[Tensor], Tensor]): The activation function in the prediction_head.
+
+    Inputs:
+        hidden_states (Tensor): The hidden states of preceding tokens.
+
+    Return:
+        Tensor: Prediction scores for the following token.
+    """
+
+    def __init__(
+        self,
+        vocab_size: int = 30522,
+        hidden_size: int = 768,
+        layer_norm_eps: float = 1e-12,
+        transform_act_fn: Callable[[Tensor], Tensor] = nn.functional.gelu,
+    ) -> None:
+        super().__init__()
+        self.dense = nn.Linear(hidden_size, hidden_size)
+        self.transform_act_fn = transform_act_fn
+        self.layer_norm = nn.LayerNorm(hidden_size, eps=layer_norm_eps)
+        self.decoder = nn.Linear(hidden_size, vocab_size)
+
+    def forward(self, hidden_states: Tensor) -> Tensor:
+        hidden_states = self.dense(hidden_states)
+        hidden_states = self.transform_act_fn(hidden_states)
+        hidden_states = self.layer_norm(hidden_states)
+        hidden_states = self.decoder(hidden_states)
+        return hidden_states
+
+
+class ALBEFDecoder(nn.Module):
+    """
+    Generate the prediction scores for answers from image and question hidden states.
+
+    Args:
+        text_embeddings (ALBEFTextEmbeddings): Instantiated ALBEFTextEmbeddings.
+        multimodal_decoder (ALBEFMultimodalEncoder): Instantiated ALBEFMultimodalEncoder.
```

Comment:
this part is confusing to lay man. why is encoder being passed to decoder? please add some explanation in comments

Reply:
Originally there was a flag that turns the multimodal encoder to a decoder. But now the functionality is refactored out from the class. I renamed the variable name to multimodal_encoder to avoid confusion.